src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs

```diff
@@ -0,0 +1,388 @@
+        public static SymbolTable NewFromRecord(RecordType type)
+        {
+            return new SymbolTableOverRecordType(type ?? RecordType.Empty());
+        }
+
+        // $$$ Should NameLookupInfo impl IExternalEntity? 
+        // Merge with _environmentSymbols? 
+        private readonly Dictionary<string, NameLookupInfo> _variables = new Dictionary<string, NameLookupInfo>();
+
+        // $$$ Dataverse will set this. 
+        internal readonly IExternalDocument _document;
+
+        private readonly List<TexlFunction> _functions = new List<TexlFunction>();
+
+        // Which enums are available. 
+        // These do not compose - only bottom one wins. 
+        private EnumStoreBuilder _enumStoreBuilder;
+        private EnumSymbol[] _enumSymbolCache;
+
+        private EnumSymbol[] GetEnumSymbolSnapshot
+        {
+            get
+            {
+                // $$$ Called may add to the builder after we've assigned. 
+                // So delay snapshot until we actually need to read it. 
+                if (_enumStoreBuilder == null)
+                {
+                    _enumSymbolCache = new EnumSymbol[] { };
+                }
+
+                if (_enumSymbolCache == null)
+                {
+                    _enumSymbolCache = _enumStoreBuilder.Build().EnumSymbols.ToArray();
+                }
+
+                return _enumSymbolCache;
+            }
+        }
+
+        internal EnumStoreBuilder EnumStoreBuilder
+        {
+            get => _enumStoreBuilder;
+            set
+            {
+                _version++;
+                _enumStoreBuilder = value;
+            }
+        }
+
+        internal IEnumerable<TexlFunction> Functions => _functions;
+
+        IExternalDocument INameResolver.Document => _document;
+
+        IExternalEntityScope INameResolver.EntityScope => throw new NotImplementedException();
+
+        IExternalEntity INameResolver.CurrentEntity => default;
+
+        DName INameResolver.CurrentProperty => default;
+
+        DPath INameResolver.CurrentEntityPath => default;
+
+        bool INameResolver.SuggestUnqualifiedEnums => false;
+
+        IEnumerable<TexlFunction> INameResolver.Functions => _functions;
+
+        internal readonly Dictionary<DName, IExternalEntity> _environmentSymbols = new Dictionary<DName, IExternalEntity>();
+
+        internal DisplayNameProvider _environmentSymbolDisplayNameProvider = new SingleSourceDisplayNameProvider();
+                
+        IReadOnlyDictionary<string, NameLookupInfo> IGlobalSymbolNameResolver.GlobalSymbols => _variables;
+        
+        /// <summary>
+        /// Provide variable for binding only.
+        /// Value must be provided at runtime via <see cref="RuntimeConfig.SymbolValues"/>.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="type"></param>
+        public void AddVariable(string name, FormulaType type)
+        {
+            _version++;
```

Comment:
> _version++; [](http://example.com/codeflow?start=11&length=12)

unsafe if two thread are calling this in parallel
Should be in GetNextVersion() where we do the needful

Reply:
not a thread-safe collection, by-design.